File: src/sort/mergeSort.ts
export default function mergeSort(arr: Array<number>): Array<number> {
  if (arr.length < 2) return arr;

  const length = arr.length;
  const middle = Math.floor(length / 2);
  const left = arr.slice(0, middle);
  const right = arr.slice(middle);

  const sortedLeft = mergeSort(left);
  const sortedRight = mergeSort(right);

  return merge(sortedLeft, sortedRight);
}

function merge(
  sortedArr1: Array<number>,
  sortedArr2: Array<number>
): Array<number> {
  const result: Array<number> = [];

  while (sortedArr1.length && sortedArr2.length) {
    if (sortedArr1[0] <= sortedArr2[0]) {
      result.push(sortedArr1.shift() as number);
    } else {
      result.push(sortedArr2.shift() as number);
    }
  }

  return result.concat(sortedArr1, sortedArr2);
}

File: src/sort/mergeSort.go
package sort

func MergeSort(arr []int) []int {
	if len(arr) < 2 {
		return arr
	}

	middle := len(arr) / 2
	left := MergeSort(arr[0:middle])
	right := MergeSort(arr[middle:])

	return sort(left, right)
}

func sort(left, right []int) []int {
	result := []int{}

	for len(left) > 0 && len(right) > 0 {
		if left[0] < right[0] {
			result = append(result, left[0])

			if len(left) < 2 {
				left = left[0:0]
			} else {
				left = left[1:]
			}
		} else {
			result = append(result, right[0])
			if len(right) < 2 {
				right = right[0:0]
			} else {
				right = right[1:]
			}
		}

	}

	for _, num := range left {
		result = append(result, num)
	}

	for _, num := range right {
		result = append(result, num)
	}

	return result
}
